vxvas2nets/vas2nets_ussd.py

```diff
@@ -0,0 +1,223 @@
+            transport_metadata={
+                'vas2nets_ussd': {
+                    'sessionid': values['sessionid'],
+                }
+            })
+
+    @inlineCallbacks
+    def handle_outbound_message(self, message):
+        self.emit("Vas2NetsUssdTransport consuming %r" % (message,))
+
+        missing_fields = self.ensure_message_values(
+            message, ['in_reply_to', 'content', 'to_addr'])
+        if missing_fields:
+            nack = yield self.reject_message(message, missing_fields)
+            returnValue(nack)
+
+        endofsession = (
+            message["session_event"] == TransportUserMessage.SESSION_CLOSE)
+        if endofsession is True:
+            yield self.session_manager.clear_session(
+                message['transport_metadata']['vas2nets_ussd']['sessionid'])
+
+        response_data = {
+            'userdata': message['content'],
+            'endofsession': endofsession,
+            'msisdn': message['to_addr'],
+        }
+
+        response_id = self.finish_request(
+            message['in_reply_to'], json.dumps(response_data))
+
+        if response_id is not None:
+            ack = yield self.publish_ack(
+                user_message_id=message['message_id'],
+                sent_message_id=message['message_id'])
+            returnValue(ack)
+        else:
+            nack = yield self.publish_nack(
+                user_message_id=message['message_id'],
+                sent_message_id=message['message_id'],
+                reason="Could not find original request.")
+            returnValue(nack)
+
+    def on_down_response_time(self, message_id, time):
+        request = self.get_request(message_id)
+        # We send different status events for error responses
+        if request.code < 200 or request.code >= 300:
+            return
+        return self.add_status(
+            component='response',
+            status='down',
+            type='very_slow_response',
+            message='Very slow response',
+            reasons=[
+                'Response took longer than %fs' % (
+                    self.response_time_down,)
+            ],
+            details={
+                'response_time': time,
+            })
+
+    def on_degraded_response_time(self, message_id, time):
+        request = self.get_request(message_id)
+        # We send different status events for error responses
+        if request.code < 200 or request.code >= 300:
+            return
+        return self.add_status(
+            component='response',
+            status='degraded',
+            type='slow_response',
+            message='Slow response',
+            reasons=[
+                'Response took longer than %fs' % (
+                    self.response_time_degraded,)
+            ],
+            details={
+                'response_time': time,
+            })
+
+    def on_good_response_time(self, message_id, time):
+        request = self.get_request(message_id)
```

Comment:
This function is called by `finish_request` inside of its `if request: ` block (in the base HttpRpc transport), before the `remove_request` call, so we are guaranteed to have a request at this point.